File: Python/108.Convert Sorted Array to Binary Search Tree.py
#根据排序数组生成一个平衡二叉排序树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        return self.toBST(nums,0,len(nums)-1)

    def toBST(self,nums,l,r):
        if l>r:
            return None
        mid = (l+r)//2
        node = TreeNode(nums[mid])
        node.left = self.toBST(nums,l,mid-1)
        node.right = self.toBST(nums,mid+1,r)
        return node

class Ssolution(object):
    def inorderTraversal(self, root):
        self.result = []
        self.inorder(root)
        return self.result

    def preorderTraversal(self, root):
        self.result = []
        self.preorder(root)
        return self.result

    def inorder(self,p):
        if p!=None:
            self.inorder(p.left) if p.left!=None else None
            self.result.append(p.val)
            self.inorder(p.right) if p.right!=None else None

    def preorder(self,p):
        if p!=None:
            self.result.append(p.val)
            self.preorder(p.left) if p.left!=None else None
            self.preorder(p.right) if p.right!=None else None

ss = Ssolution()
s = Solution()
root = s.sortedArrayToBST([1,2,3,5,8,9,15,30])
print(ss.inorderTraversal(root))
print(ss.preorderTraversal(root))

File: C++/Tree/108.Convert Sorted Array to Binary Search Tree.cpp
/*
把有序数组转换为二叉搜索树。
*/
/*
思路：把中间结点拿出来新建根结点，然后把数组下标传入递归函数分别递归左右子树。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size()==0)	return NULL;
        return createBST(nums,0,nums.size()-1);
    }
    TreeNode* createBST(vector<int>& nums,int left,int right)
    {
    	if(left>right) return NULL;
        if(left==right) return new TreeNode(left);
    	int mid = (left+right)/2;
    	TreeNode* root = new TreeNode(nums[mid]);
    	TreeNode* leftnode = createBST(nums,left,mid-1);
    	TreeNode* rightnode = createBST(nums,mid+1,right);
    	root->left = leftnode, root->right = rightnode;
    	return root;
    }
};
int main()
{
	Solution s;
	vector<int> nums{1,2,3,4,5,6,7,8};
	TreeNode* out = s.sortedArrayToBST(nums);
}

